expo/expo
# [expo-constants] Change source of truth for constants types

# Why

This fixes two things:
- [This todo](https://github.com/expo/expo/blob/main/packages/expo-updates/src/Updates.types.ts#L24)
- [This task](https://linear.app/expo/issue/ENG-9068/constantsexpoconfig-missing-keys-contains-incorrect-keys-in-release)

Closes ENG-9143.

# How

1. Move manifest definitions to expo-manifests package. This package is now the source of truth about what is in a manifest, both in native code and in typescript types.
2. Refactor expo-updates to use types from expo-manifests, thus completing the TODO
3. Refactor expo-dev-launcer to use types from expo-manifests, following the pattern from expo-updates since it does a similar thing
4. Refactor expo-constants to use types from expo-updates, expo-dev-launcher, and its own types depending on which manifest it is using. Note that this uses a technique from https://stackoverflow.com/a/63905376 which is an optional type import since expo-constants dynamically fills the manifest variables depending on which native modules are present, so we need to do something similar in typescript and this is the closest technique I could find. Still need to test that this actually works. (if it doesn't, we could consider making expo-constants depend on expo-manifests though this is messy)
5. For ExpoConstants.expoConfig getter, use the embedded `app.config` to resolve this field when running an embedded update from expo updates since embedded manifest doesn't have the config in it. This is a similar fix to https://github.com/expo/expo/pull/23118 and fixes the linked task.

# Test Plan

- Run unit test.

Manual test:
1. `et gba -n testwat8 expo-updates expo-manifests expo-updates-interface expo-constants`
2. `eas init`
3. `eas update:configure`
4. Put the code below in App.js
5. `npx expo run:ios --configuration Release`
6. See that:
    - `Constants.expoConfig` contains embedded app.config
    - `Constants.manifest` contains embedded manifest from expo-updates
    - `Constants.manifest2` is empty
    - Typescript types work for everything in vscode.

App.js:
```typescript
import Constants from 'expo-constants';
import { StatusBar } from 'expo-status-bar';
import { ScrollView, StyleSheet, Text, View } from 'react-native';

export default function App() {
  function blah(it) {
    if (!it) {
      return '';
    }
    return JSON.stringify(it);
  }

  return (
    <ScrollView style={styles.container} contentContainerStyle={{ paddingTop: 70 }}>
      <Text>Values: {blah(Constants.expoConfig)}</Text>
      <View style={{ marginTop: 20 }} />
      <Text>Values: {blah(Constants.manifest)}</Text>
      <View style={{ marginTop: 20 }} />
      <Text>Values: {blah(Constants.manifest2)}</Text>
      <StatusBar style="auto" />
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
});
```

# Checklist
